threeML/plugins/OGIP/event_polynomial.py

```diff
@@ -191,39 +189,59 @@ def unbinned_polyfit(events, grade, t_start, t_stop, exposure, initial_amplitude
     """
 
     # first do a simple amplitude fit
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
 
-    search_grid = np.logspace(-2, 4, 10)
+        search_grid = np.logspace(-2, 4, 10)
 
-    initial_guess = np.zeros(grade + 1)
+        initial_guess = np.zeros(grade + 1)
 
-    polynomial = Polynomial(initial_guess)
+        polynomial = Polynomial(initial_guess)
 
-    log_likelihood = PolyUnbinnedLogLikelihood(events,
-                                               polynomial,
-                                               t_start,
-                                               t_stop,
-                                               exposure)
+        # if there are no events then return nothing
 
-    like_grid = []
-    for amp in search_grid:
 
-        initial_guess[0] = amp
-        like_grid.append(log_likelihood(initial_guess))
 
-    initial_guess[0] = search_grid[np.argmin(like_grid)]
+        if len(events) == 0:
 
-    # Improve the solution
+            return Polynomial([0]), 0
 
+        log_likelihood = PolyUnbinnedLogLikelihood(events,
+                                                   polynomial,
+                                                   t_start,
+                                                   t_stop,
+                                                   exposure)
 
+        like_grid = []
+        for amp in search_grid:
 
-    final_estimate = scipy.optimize.fmin(log_likelihood, initial_guess,
-                                         ftol=1E-5, xtol=1E-5,
-                                         maxiter=1e6, maxfun=1E6,
-                                         disp=False)
+            initial_guess[0] = amp
+            like_grid.append(log_likelihood(initial_guess))
 
-    # Get the value for cstat at the minimum
+        initial_guess[0] = search_grid[np.argmin(like_grid)]
 
-    min_log_likelihood = log_likelihood(final_estimate)
+        # Improve the solution
+        dof = len(events) - (grade + 1)
+
+        if dof <= 2:
+            # Fit is poorly or ill-conditioned, have to reduce the number of parameters
+            while (dof < 1 and len(initial_guess) > 1):
+                initial_guess = initial_guess[:-1]
+                polynomial = Polynomial(initial_guess)
+                log_likelihood = PolyUnbinnedLogLikelihood(events,
+                                                           polynomial,
+                                                           t_start,
+                                                           t_stop,
+                                                           exposure)
+
+        final_estimate = scipy.optimize.fmin(log_likelihood, initial_guess,
```

Comment:
I prefer to use the scipy.optimize.minimize function, so that you can change easily algorithm, instead of one of the more specialized versions like fmin.